src/relations/hacluster.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""hacluster relation hooks and helpers."""
+
+import json
+import logging
+from hashlib import shake_128
+from ipaddress import IPv4Address, ip_address
+from typing import Optional, Union
+
+import ops
+
+HACLUSTER_RELATION_NAME = "ha"
+
+logger = logging.getLogger(__name__)
+
+
+class HACluster(ops.Object):
+    """Defines hacluster functunality."""
+
+    def __init__(self, charm: ops.CharmBase):
+        super().__init__(charm, HACLUSTER_RELATION_NAME)
+
+        self.charm = charm
+
+        self.framework.observe(
+            charm.on[HACLUSTER_RELATION_NAME].relation_changed, self._on_changed
+        )
+        self.framework.observe(charm.on.config_changed, self._on_changed)
+
+    @property
+    def relation(self) -> ops.Relation:
+        """Returns the relations in this model, or None if hacluster is not initialised."""
+        return self.charm.model.get_relation(HACLUSTER_RELATION_NAME)
+
+    def _is_clustered(self) -> bool:
+        for key, value in self.relation.data.items():
+            if isinstance(key, ops.Unit) and key != self.charm.unit:
+                if value.get("clustered") in ("yes", "true"):
+                    return True
+                break
+        return False
```

Comment:
the underlying hacluster application needs at least 3 nodes/units to be active. this helper is checking if hacluster is clustered:

```
  - relation-id: 12                                                                                                                                                                                         
    endpoint: ha                                                                                                                                                                                            
    related-endpoint: ha                                                                                                                                                                                    
    application-data: {}                                                                                                                                                                                    
    related-units:                                                                                                                                                                                          
      hacluster/0:                                                                                                                                                                                          
        in-scope: true                                                                                                                                                                                      
        data:                                                                                                                                                                                               
          clustered: "yes"                                                                                                                                                                                  
          egress-subnets: 10.205.193.104/32                                                                                                                                                                 
          ingress-address: 10.205.193.104                                                                                                                                                                   
          private-address: 10.205.193.104                                                                                                                                                                   
      hacluster/1:                                                                                                                                                                                          
        in-scope: true                                                                                                                                                                                      
        data:                                                                                                                                                                                               
          clustered: "yes"                                                                                                                                                                                  
          egress-subnets: 10.205.193.134/32                                                                                                                                                                 
          ingress-address: 10.205.193.134                                                                                                                                                                   
          private-address: 10.205.193.134                                                                                                                                                                   
      hacluster/3:                                                                                                                                                                                          
        in-scope: true                                                                                                                                                                                      
        data:                                                                                                                                                                                               
          clustered: "yes"                                                                                                                                                                                  
          egress-subnets: 10.205.193.179/32                                                                                                                                                                 
          ingress-address: 10.205.193.179                                                                                                                                                                   
          private-address: 10.205.193.179                                                          
      hacluster/4:                                                                                 
        in-scope: true                                                                             
        data:                                                                                      
          clustered: "yes"                                                                         
          egress-subnets: 10.205.193.206/32                                                        
          ingress-address: 10.205.193.206                                                          
          private-address: 10.205.193.206
```

Reply:
hacluster is clustered when there are 3 or more nodes in pacemaker/corosync. there cannot be a cluster with one unit, and if this is the case, we are not asking hacluster to set up the VIP in any of the charm's nodes. hacluster also does [not clean up resources](https://bugs.launchpad.net/charm-hacluster/+bug/2080446) when the relation between the principal and subordinate charm is removed

also, we are not really "using" hacluster, just requesting that hacluster create and manage some resources (VIP) for us

Reply:
in reality it is indeed checking if there are 3+ units, but it is more abstracted than that. if tomorrow, hacluster charm redefines how it is clustered, it can freely do so. all it needs to do is write to the databag of one unit that `clustered: yes/true`

in 9649f9edc4473383a057560f5107845cb3b8d18f, updated the implementation to accept `clustered: yes` from any related unit